### PR TITLE
NSM: provide minimum and maximum time-to-live to a netflow log

### PR DESCRIPTION
**Pull request by macpas**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [ x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x ] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
-The goal is to add extra information for network security monitoring:
- in case of netflow logging, add TTL fields (min_ttl and  max_ttl).

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

File: src/flow-util.c [hunk 1]
```diff
@@ -137,10 +137,12 @@ void FlowInit(Flow *f, const Packet *p)
     if (PKT_IS_IPV4(p)) {
         FLOW_SET_IPV4_SRC_ADDR_FROM_PACKET(p, &f->src);
         FLOW_SET_IPV4_DST_ADDR_FROM_PACKET(p, &f->dst);
+        FLOW_SET_IPV4_TTL_FROM_PACKET(p, f);
```
Comment (victorjulien) [inline src/flow-util.c]: using this macro here is a bit confusing and inefficient too, can just be:
f->min_ttl_toserver = f->max_ttl_toserver = IPV4_GET_IPTTL((p));
Don't think the macro adds anything.

File: src/flow-util.c [hunk 2]
```diff
@@ -137,10 +137,12 @@ void FlowInit(Flow *f, const Packet *p)
...
         f->flags |= FLOW_IPV4;
     } else if (PKT_IS_IPV6(p)) {
         FLOW_SET_IPV6_SRC_ADDR_FROM_PACKET(p, &f->src);
         FLOW_SET_IPV6_DST_ADDR_FROM_PACKET(p, &f->dst);
+        FLOW_SET_IPV6_HLIM_FROM_PACKET(p, f);
```
Comment (victorjulien) [inline src/flow-util.c]: same here

File: src/flow.c [hunk 3]
```diff
@@ -266,6 +266,29 @@ static inline int FlowUpdateSeenFlag(const Packet *p)
...
+{
+    if (FlowGetPacketDirection(f, p) == TOSERVER) {
+        if (ttl < f->min_ttl_toserver) {
+            f->min_ttl_toserver = ttl;
+        } else if (f->min_ttl_toserver == 0) {
+            f->min_ttl_toserver = ttl;
+        }
+        if (ttl > f->max_ttl_toserver) {
+            f->max_ttl_toserver = ttl;
+        }
```
Comment (victorjulien) [inline src/flow.c]: How about compressing this to:
<pre>
if (f->min_ttl_toserver == 0) {
    f->min_ttl_toserver = ttl;
} else {
   f->min_ttl_toserver = MIN(f->min_ttl_toserver, ttl);
}
f->max_ttl_toserver = MAX(f->max_ttl_toserver, ttl);
</pre>

File: src/flow.c [hunk 4]
```diff
@@ -344,6 +367,16 @@ void FlowHandlePacketUpdate(Flow *f, Packet *p)
         SCLogDebug("setting FLOW_NOPAYLOAD_INSPECTION flag on flow %p", f);
         DecodeSetNoPayloadInspectionFlag(p);
     }
+
+
+    /* update flow's ttl fields if needed */
+    if (PKT_IS_IPV4(p)) {
+        uint8_t ttl = IPV4_GET_IPTTL(p);
```
Comment (victorjulien) [inline src/flow.c]: can be a single line: FlowUpdateTTL(f, p, IPV4_GET_IPTTL(p));

File: src/flow.h [hunk 5]
```diff
@@ -183,6 +183,16 @@ typedef struct AppLayerParserState_ AppLayerParserState;
         (a)->addr_data32[3] = (p)->ip6h->s_ip6_dst[3];  \
     } while (0)
 
+#define FLOW_SET_IPV4_TTL_FROM_PACKET(p, f) do {    \
```
Comment (victorjulien) [inline src/flow.h]: I'm not in favor of these, in part because it confused me when I first looked why these only set toserver

File: src/flow.h [hunk 6]
```diff
@@ -330,6 +340,10 @@ typedef struct Flow_
     };
     uint8_t proto;
     uint8_t recursion_level;
+    uint8_t min_ttl_toserver;
```
Comment (victorjulien) [inline src/flow.h]: please move below vlan_id so all the fields used for hashing are together

File: src/output-json-netflow.c [hunk 7]
```diff
@@ -263,6 +266,12 @@ static void JsonNetFlowLogJSONToClient(JsonNetFlowLogThread *aft, json_t *js, Fl
     json_object_set_new(hjs, "age",
             json_integer(age));
 
+    /* To client is zero if we did not see any packet */
+    if (f->max_ttl_toclient) {
```
Comment (victorjulien) [inline src/output-json-netflow.c]: f->tosrcpktcnt seems like a better test

File: src/flow-util.c (hunk 1 above)
Comment (macpas) [reply src/flow-util.c]: I do agree.

File: src/flow-util.c (hunk 2 above)
Comment (macpas) [reply src/flow-util.c]: I do agree.

File: src/flow.c (hunk 3 above)
Comment (macpas) [reply src/flow.c]: OK and same I suppose for the next piece of code (toclient).

File: src/flow.c (hunk 4 above)
Comment (macpas) [reply src/flow.c]: OK it is better with one single line.

File: src/flow.h (hunk 5 above)
Comment (macpas) [reply src/flow.h]: Yes, it can be removed now.

File: src/flow.h (hunk 6 above)
Comment (macpas) [reply src/flow.h]: OK I do it, TTL variables are moved below vlan_id.

File: src/output-json-netflow.c (hunk 7 above)
Comment (macpas) [reply src/output-json-netflow.c]: OK it seems OK.

**Comment by victorjulien** [comment]
Who is going to implement these changes? If @regit or @glongo they can update their existing commits. If @macpas then I think they should be implemented in an additional commit.

In any case please submit a new PR with a new branch. See https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Github_work_flow#Create-a-new-branch-for-incorporating-feedback

Thanks!

**Pull request closed without merging**